src/main/java/com/compilercharisma/chameleonbusinessstudio/controller/AppointmentController.java

```diff
@@ -145,34 +148,109 @@ public Mono<ResponseEntity<?>> update(@RequestBody AppointmentEntity appointment
         }
 
         var postedBy = authentication.getLoggedInUser().getAsEntity();
         if(!appointments.isUserRegistered(appointment, postedBy.getEmail())){
             appointments.registerUser(appointment, postedBy.getEmail());
         }
 
-        return Mono.just(ResponseEntity.ok().body(appointment));
+        return Mono.just(ResponseEntity.ok(appointment));
+    }
+
+    /**
+     * POST /api/v1/appointments/123?email=foo.bar@baz.qux
+     * Books a user for an appointment.
+     * 
+     * @param appointmentId the ID of the appointment to book the given email 
+     *  for
+     * @param email the email of the user to book an appointment for
+     * @return either an OK response with the updated appointment as its body, 
+     *  or one of several error responses
+     */
+    @PostMapping("/book-them/{appointment-id}")
+    public Mono<ResponseEntity<AppointmentEntity>> bookThem(
+            @PathVariable("appointment-id") int appointmentId,
+            @RequestParam("email") String email
+    ){
+        if(email == null){
+            return Mono.error(new IllegalArgumentException("Email cannot be null"));
+        }
+        if(!isValidEmail(email)){
+            return Mono.error(new IllegalArgumentException("Invalid email: " + email));
+        }
+        if(!users.isRegistered(email)){
+            return Mono.error(new IllegalArgumentException("Email is not registered as a user: " + email));
+        }
+
+        var appointmentMaybe = appointments.getAppointmentById(appointmentId);
+        if(appointmentMaybe.isEmpty()){
+            return Mono.error(new IllegalArgumentException("Invalid appointment ID: " + appointmentId));
+        }
+
+        // check if logged in user is allowed to book
+        var postedBy = authentication.getLoggedInUser().getAsEntity();
+        if(!isRoleAllowedToBookOtherPeople(postedBy.getRole())){
+            var msg = String.format(
+                "Users with role \"%s\" are not allowed to book appointments for other users",
+                postedBy.getRole()
+            );
+            return Mono.error(new AccessDeniedException(msg));
+        }
+        
+        var appointment = appointmentMaybe.get();
+        if(!appointments.isSlotAvailable(appointment)){
+            return Mono.error(new UnsupportedOperationException("Appointment is full; it cannot have any more participants"));
+        }
+
+        if(!appointments.isUserRegistered(appointment, email)){
+            appointments.registerUser(appointment, email);
+        }
+
+        return Mono.just(ResponseEntity.ok(appointment));
     }
     
     private boolean isRoleAllowedToCreateAppointments(String role){
-        HashSet<String> hs = new HashSet<>();
+        var hs = new HashSet<>();
         hs.add(Role.ADMIN);
         hs.add(Role.ORGANIZER);
         hs.add(Role.TALENT);
         return hs.contains(role);
```

Comment:
You don't need this, because how we will handle roles and authorization will be handled through Spring Security and Authorization Managers.

Reply:
We need this since users are logged in as `OAuth2` users, and our roles are attached to our domain representation of users, not the `OAuth2` ones. **If** I have time, I'd like to set the Principal to our user, but that's a big **if**.